### PR TITLE
Cleanups for `<deque>` part 1

### PR DESCRIPTION
**Pull request by achabense**
Contents of this pr:
1. `trivial relocations XXX`: The definition order of `deque`'s methods was highly chaotic; reorder them to roughly match with the standard.
2. `A&B`: some minor cleanups for push/insert functions
3. `C`: Let `_Construct[_n]` do proxy allocation to save some lines.
4. ~`D`: The usage of `_Allocate_at_least_helper` in `_Growmap` was problematic, as it can make `_Mapsize()` not power of 2 if `_Allocate_at_least_helper` changes `_Newsize`; restore to the old version.~ moved to pr 4017

### CODE REVIEW

**Comment by achabense** [comment]
Ugh, I hit `enter` by mistake

File: stl/inc/deque
Comment (achabense) [inline stl/inc/deque:1524]: This reverts the change introduced by #3864; do I need to add test for it?

**Comment by achabense** [comment]
https://github.com/microsoft/STL/blob/d8ed02d2ec7ba8216848a33dec2b5fb385febf4e/tests/std/tests/GH_003570_allocate_at_least/test.cpp#L74-L79
Can I just change to `!assert(allocate_at_least_signal.consume());` here?

**Comment by achabense** [comment]
The bug is nastiy one; I will move it to a separate pr...

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! The fine-grained commits, combined with VSCode's new ability to identify moved code, made this easy to review. Everything looks good and I didn't find any leftover redundant access specifiers.

**Comment by StephanTLavavej** [comment]
Despite the magnitude of the edits, the well-structured fine-grained commits made this very easy to review, and I am confident that we don't need a second maintainer approval.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for making `<deque>` match the Standard more obviously! :tada: :smile_cat: :rocket:

**Pull request merged**
